libraries/chain/controller.cpp

```diff
@@ -3994,48 +4009,45 @@ std::optional<signed_block_header> controller::fetch_block_header_by_id( const b
 }
 
 signed_block_ptr controller::fetch_block_by_number( uint32_t block_num )const  { try {
-   auto blk_state = fetch_block_state_by_number( block_num );
-   if( blk_state ) {
-      return blk_state->block;
-   }
+   auto fetch_block = [&](auto& fork_db, auto& head) -> signed_block_ptr {
+      auto bsp = fork_db.search_on_branch( fork_db.head()->id(), block_num);
+      if (bsp) return bsp->block;
+      return {};
+   };
+   auto b = my->block_data.apply<signed_block_ptr>(fetch_block);
+   if (b)
+      return b;
 
    return my->blog.read_block_by_num(block_num);
 } FC_CAPTURE_AND_RETHROW( (block_num) ) }
 
 std::optional<signed_block_header> controller::fetch_block_header_by_number( uint32_t block_num )const  { try {
-   auto blk_state = fetch_block_state_by_number( block_num );
-   if( blk_state ) {
-      return blk_state->header;
-   }
+   auto fetch_block = [&](auto& fork_db, auto& head) -> signed_block_ptr {
+      auto bsp = fork_db.search_on_branch( fork_db.head()->id(), block_num);
+      if (bsp) return bsp->block;
+      return {};
+   };
+   auto b = my->block_data.apply<signed_block_ptr>(fetch_block);
+   if (b)
+      return *b;
 
    return my->blog.read_block_header_by_num(block_num);
 } FC_CAPTURE_AND_RETHROW( (block_num) ) }
 
 
-block_state_legacy_ptr controller::fetch_block_state_by_id( block_id_type id )const {
-   // returns nullptr when in IF mode 
-   auto get_block_state = [&](auto& fork_db, auto& head) -> block_state_legacy_ptr { return fork_db.get_block(id); };
-   return my->block_data.apply_dpos<block_state_legacy_ptr>(get_block_state);
-}
-
-block_state_legacy_ptr controller::fetch_block_state_by_number( uint32_t block_num )const  {
-   try {
-      // returns nullptr when in IF mode
-      auto fetch_block_state = [&](auto& fork_db, auto& head) -> block_state_legacy_ptr {
-         return fork_db.search_on_branch( fork_db.head()->id(), block_num);
-      };
-      return  my->block_data.apply_dpos<block_state_legacy_ptr>(fetch_block_state);
-   } FC_CAPTURE_AND_RETHROW( (block_num) )
-}
-
 block_id_type controller::get_block_id_for_num( uint32_t block_num )const { try {
    const auto& blog_head = my->blog.head();
 
    bool find_in_blog = (blog_head && block_num <= blog_head->block_num());
 
    if( !find_in_blog ) {
-      auto bsp = fetch_block_state_by_number( block_num );
-      if( bsp ) return bsp->id();
+      auto fetch_block_id = [&](auto& fork_db, auto& head) -> std::optional<block_id_type> {
+         auto bsp = fork_db.search_on_branch( fork_db.head()->id(), block_num);
+         if (bsp) return bsp->id();
+         return {};
+      };
+      auto id = my->block_data.apply<std::optional<block_id_type>>(fetch_block_id);
```

Comment:
Same here, could define in `struct block_data_t`:

```
   block_id_type fetch_block_id(uint32_t block_num) const {
      return std::visit([&](const auto& bd) -> block_id_type {
         auto bsp = bd.fork_db.search_on_branch(bd.fork_db.head()->id(), block_num);
         if (bsp) return bsp->id();
         return {};
      }, v);
   }
```

Reply:
Done.